### PR TITLE
Task 947

### LINKED ISSUE #944: As a sysadmin, I want to quantify the shortest running region for which the EE agent does not lead to incorrect freq assignment of consecutive regions
## Problem to be identified: 
The frequency assigned by the EE-agent on a short-running region creeps over to the next adjacent region. 

## Story goal:
The frequency map agent to assign different frequencies to adjacent regions, both consuming the same power, but one whose duration is frequency-dependent and the other, independent. The assumption here is that these regions will run over multiple iterations

## Done criteria: 
A plot w/ X-axis = region duration, and Y-axis = energy savings v/s static sticker. Secondary Y-axis = runtime degration v/s static sticker assignment

### LINKED ISSUE #947: Write test application and python script that runs timed-scaling region and the scaling region with different big-o's
> estimate 5

The script should execute the benchmark twice, once with the frequency map agent pegged to sticker, and the other with the frequency map agent set to alternate between min for the timed-scaling region and sticker for the scaling region.  Note that for each loop, the execution time is constant, but more regions are executed if they are shorter.

Done criteria:
Make plot described in #944.

Parent story #944

### PR DESCRIPTION
**Pull request by cmcantalupo**
Fixes #947
Fixes #944

Depends on #959

### CODE REVIEW

File: test_integration/test_ee_short_region_slop.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            geopm::ModelRegion::model_region("scaling", duration, is_verbose));
+        std::unique_ptr<geopm::ModelRegion> model_timed(
+            geopm::ModelRegion::model_region("timed_scaling", duration, is_verbose));
+
+        // Rename model regions
+        std::string scaling_name = "scaling_" + std::to_string(trial_idx);
+        std::string timed_name = "timed_" + std::to_string(trial_idx);
+        uint64_t scaling_rid = prof.region(scaling_name, GEOPM_REGION_HINT_UNKNOWN);
+        uint64_t timed_rid = prof.region(timed_name, GEOPM_REGION_HINT_UNKNOWN);
+        // Execute the regions back to back repeatedlythe renamed region
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.cpp]: Clean up kruft at the end of comment

File: test_integration/test_ee_short_region_slop.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                is_verbose = true;
+            }
+        }
+    }
+    geopm::Profile &prof = geopm::Profile::default_profile();
+    // Run twelve trials with region duration ranging from 100 us - 400 ms
+    size_t num_trial = 12;
+    double duration = 1e-4;
+    double repeat = 409600; // Each trial takes 41 seconds and the
+			    // whole execution takes 8 minutes
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.cpp]: 16 minutes at sticker frequency.

File: test_integration/test_ee_short_region_slop.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            sys.exc_clear()
+        if not cls._skip_launch:
+            # Set the job size parameters
+            num_node = 1
+            num_rank = 1
+            time_limit = 6000
+            # Configure the test application
+            app_conf = AppConf()
+
+            # Region hases for scaling_0, scaling_1, ... , scaling_11
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: hashes

File: test_integration/test_ee_short_region_slop.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+            scaling_rid = [0x8a244fc3, 0xc31832e4, 0x185cb58d, 0x5160c8aa,
+                           0xab39cdae, 0xe205b089, 0x394137e0, 0x707d4ac7,
+                           0xc81f4b19, 0x8123363e, 0x62184bff, 0x0a1b6737]
+            # Region hases for timed_0, timed_1, ... , timed_11
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: hashes

File: test_integration/test_ee_short_region_slop.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+            timed_rid = [0x2dbeb83f, 0x3e1c2048, 0x0afb88d1, 0x195910a6,
+                         0x6334d9e3, 0x70964194, 0x4471e90d, 0x57d3717a,
+                         0xb0aa7b87, 0xa308e3f0, 0x0eff69f9, 0xfc94eafa]
+            # Configure the agent
+            # Query for the min and sticker frequency and run the
+            # energy efficient agent over this range.
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: frequency map agent

File: test_integration/test_ee_short_region_slop.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            freq_min = geopm_test_launcher.geopmread("CPUINFO::FREQ_MIN board 0")
+            freq_sticker = geopm_test_launcher.geopmread("CPUINFO::FREQ_STICKER board 0")
+            agent_conf_fixed_dict = {'FREQ_MIN':freq_min,
+                               'FREQ_MAX':freq_sticker}
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: whitespace

File: test_integration/test_ee_short_region_slop.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        cols, scaling_data_dynamic, timed_data_dynamic = extract_data(report_dynamic)
+
+        report_fixed = geopmpy.io.RawReport(self._report_path_fixed)
+        cols, scaling_data_fixed, timed_data_fixed = extract_data(report_fixed)
+
+        all_energy = list(scaling_data_fixed[column_idx(cols, 'package-energy')])
+        all_energy.extend(scaling_data_dynamic[column_idx(cols, 'package-energy')])
+        all_energy.extend(timed_data_fixed[column_idx(cols, 'package-energy')])
+        all_energy.extend(timed_data_dynamic[column_idx(cols, 'package-energy')])
+        min_energy = min(all_energy)
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py:208]: Add padding to the image

File: test_integration/test_ee_short_region_slop.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        max_energy = max(all_energy)
+
+        all_runtime = list(scaling_data_fixed[column_idx(cols, 'runtime')])
+        all_runtime.extend(scaling_data_dynamic[column_idx(cols, 'runtime')])
+        all_runtime.extend(timed_data_fixed[column_idx(cols, 'runtime')])
+        all_runtime.extend(timed_data_dynamic[column_idx(cols, 'runtime')])
+        min_runtime = min(all_runtime)
+        max_runtime = max(all_runtime)
+
+        plt.figure(figsize=(10,10))
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: increase the figure size or change units to kjoules

File: test_integration/test_ee_short_region_slop.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    cols = [('count', ''),
+            ('package-energy', 'joules'),
+            ('requested-online-frequency', ''),
+            ('power', 'watts'),
+            ('runtime', 'sec'),
+            ('frequency', 'Hz')]
+    scaling_data = extract_data_region(report, cols, 'scaling')
+    timed_data = extract_data_region(report, cols, 'timed')
+    cols.append(('duration', 'sec'))
+    dur = tuple(rt / ct for ct, rt in zip(scaling_data[0], scaling_data[4]))
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: use column_idx() function not magic number

File: test_integration/test_ee_short_region_slop.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    scaling_data.append(dur)
+    dur = tuple(rt / ct for ct, rt in zip(timed_data[0], timed_data[4]))
```
Comment (cmcantalupo) [inline test_integration/test_ee_short_region_slop.py]: same as above

**Review by cmcantalupo — COMMENTED** [review]
Would be nice to have more comments in the code too.

File: test_integration/test_ee_short_region_slop.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+"""EE_SHORT_REGION_SLOP
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py:34]: ?

File: test_integration/test_ee_short_region_slop.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+        all_runtime.extend(scaling_data_dynamic[column_idx(cols, 'runtime')])
+        all_runtime.extend(timed_data_fixed[column_idx(cols, 'runtime')])
+        all_runtime.extend(timed_data_dynamic[column_idx(cols, 'runtime')])
+        min_runtime = min(all_runtime)
+        max_runtime = max(all_runtime)
+        delta_runtime = max_runtime - min_runtime
+        min_runtime -= 0.05 * delta_runtime
+        max_runtime += 0.05 * delta_runtime
+
+        plt.figure(figsize=(11,11))
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py:224]: This code that uses global plt directly could instead use the objects returned by plt.subplots()

File: test_integration/test_ee_short_region_slop.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+    else:
+        plt.xlabel(xaxis)
+    units = cols[yaxis_idx][1]
+    if units:
+        plt.ylabel('{} ({})'.format(yaxis, units))
+    else:
+        plt.ylabel(yaxis)
+
+def extract_data_region(report, cols, key):
+    host = report.host_names()[0]
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py:291]: could pass in host as a parameter?

**Review by dianarg — CHANGES_REQUESTED** [review]

File: test_integration/test_ee_short_region_slop.py (hunk 11 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py:34]: Added a long description

File: test_integration/test_ee_short_region_slop.py (hunk 12 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py:224]: Maybe next time?

File: test_integration/test_ee_short_region_slop.py (hunk 13 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py:291]: Could do, and if we promoted this implementation to a separate object that would make sense.  In this case the test is only run on one host.

**Review by dianarg — APPROVED** [review]

**Pull request merged**